README.md
#weekly

jQuery Weekly Calendar Plugin

##Development

###Requirements

- node and npm
- bower `npm install -g bower`
- grunt `npm install -g grunt-cli`

###Setup

- `npm install`
- `bower install`

###Run

`grunt dev`

or for just running tests on file changes:

`grunt ci`

###Tests

`grunt mocha`


HISTORY.md

0.0.38 / 2013-10-08 
==================

  * added setReadOnly method
  * updated grunt to just run default on watch
  * few changes to the build scripts
  * updated to use load-grunt-config

0.0.37 / 2013-10-02 
==================

  * added setSplitInterval method

0.0.36 / 2013-10-02 
==================

  * added property to set autoSplitInterval independant from interval
  * updated example to toggle autosplit
  * method to setAutoSplit

0.0.35 / 2013-09-27 
==================

  * fixed split event to support interval
  * Added ability to pass in interval.

0.0.34 / 2013-09-26 
==================

  * tweaked layout of delete button
  * added livereload to grunt-connect
  * Killed plato (hehe).
  * Updated tests files.
  * Added test for fromDecimal
  * Allows events to be created in 30 minute intervals.
  * Tweaked time border colors.
  * Added half hour guides.
  * load-grunt-tasks and bower ignore

0.0.33 / 2013-09-05 
==================

  * Fixed isPastDate

0.0.32 / 2013-08-29 
==================

  * moved livereload option to the top
  * don't return time when calling getDates()
  * renamed getdateUtils to getDates

0.0.31 / 2013-08-26 
==================

  * fixed tests
  * Added support to start the week on any given week offset (0-6) or today.

0.0.30 / 2013-08-21 
==================

  * Fixed issue when mouseDown would be set too early. Also added check for clicking on close button.

0.0.29 / 2013-08-20 
==================

  * Merge branch 'feature/create-previous-events'
  * updated build
  * Added option to disable/enable creating past events.

0.0.28 / 2013-08-19 
==================

  * updated to use bower 1.0

0.0.27 / 2013-08-09 
==================

  * updated timezone should not trigger weekchange
  * [grunt] don't run default on dev
  * addEvent shouldn't offset time

0.0.26 / 2013-08-07 
==================

  * Merge remote-tracking branch 'origin/bug/time-display-midnight'
  * Merge remote-tracking branch 'origin/feature/timezoneOffset'
  * removed moment-timezone from exclude
  * Fixed problem with midnight display.
  * Fixed issue with new events being offset when they didn't need to be.

0.0.25 / 2013-08-06 
==================

  * Merge branch 'feature/timezoneOffset'
  * Updated setTimezoneOffset to actually work correctly and not just offset everything.
  * added timezone dropdown to example
  * First version of timezone support.

0.0.24 / 2013-08-05 
==================

  * grunt connect to allow connections from anywhere
  * tweaks to ui to fix scrollbar offset issue
  * fixed display in firefox
  * updated style for delete x

0.0.23 / 2013-08-02 
==================

  * Merge remote-tracking branch 'origin/bug/scheduling-last-time'
  * Fixed issue with months being one off.
  * Fixed rendering bug if event ends at midnight.

0.0.22 / 2013-07-31 
==================

  * fixed start time of 0

0.0.21 / 2013-07-28 
==================

  * Merge branch 'feature/scroll-times'
  * added startTimeScrollOffset option
  * [example] updated to end at midnight
  * [grunt] run default on dev
  * Updated example.
  * Added scroller around times and grid with option in theme mixin.

0.0.20 / 2013-07-22 
==================

  * add event to events[] even if outside bounds

0.0.19 / 2013-07-22 
==================

  * fixed for including the hour after the endTime

0.0.18 / 2013-07-22 
==================

  * don't render times outside of startTime and endTime
  * added showToday option to not show today button

0.0.17 / 2013-07-17 
==================

  * added border around calendar, added option for eventSpacing

0.0.16 / 2013-07-17 
==================

  * fixed delete button location

0.0.15 / 2013-07-16 
==================

  * option to allowPreviousWeeks which hides the previous week button
  * make delete button bigger

0.0.14 / 2013-07-12 
==================

  * if event has a type attribute, add weekly-event-[type] class
  * test for populating title
  * changed name to title
  * added showing of event title

0.0.13 / 2013-07-09 
==================

  * added time.js to build
  * removed reloadr, added livereload option to watch

0.0.12 / 2013-07-08 
==================

  * added event trigger when clicking on an event
  * added enableDelete.  if readOnly, set enableDelete and enableResize to false
  * Renamed Dates to dateUtil.
  * Added Dates lib and tests. Updated weekly to reflect these changes.
  * Removed time lib from dist directory.
  * Changed weekly over to use new TimeFormat library.
  * Added TimeFormat library & tests.

0.0.11 / 2013-07-01 
==================

  * option to disable resize
  * added autoSplit option to automatically split events into hour intervals

0.0.10 / 2013-06-24 
==================

  * Added readonly mode + test.
  * Basic test for draf to modify.
  * Finished drag to modify event.

0.0.9 / 2013-06-20 
==================

  * fixed issue where tests weren't running on 2nd save of test files
  * added weekChange event

0.0.8 / 2013-06-18 
==================

  * updated example
  * changed clearedEvents to clearEvents
  * changed .id to ._index so .id can be used for fetching from db

0.0.7 / 2013-06-18 
==================

  * set background color for weekly-dragger based on eventColor
  * option for min/max for fitText

0.0.6 / 2013-06-18 
==================

  * WIP - Drag to modify event.
  * Font sizes are a little more reasonable.
  * Added dates to next/previous buttons.
  * Event time now wraps.
  * Updated header week on prev/next.

0.0.5 / 2013-06-11 
==================

  * update to user concat-bower

0.0.4 / 2013-06-11 
==================

  * Hides today button when on today + tests.
  * Added jump to today method and added test.
  * Changed changeWeek methods to use weekOffset.
  * Actually fized timezone offset test.
  * Fixed breaking time test.
  * Fixed breaking tests due to fittext.
  * Fixed bug where event would be created when right or middle clicking.

test/time.test.js

suite('TimeFormat', function() {

  suite('time format', function() {
    test('day', function() {
      var date = new Date(2013, 4, 8);

      assert.equal(TimeFormat('%d %j %D %l', date), '08 8 Wed Wednesday');
    });

    test('month', function() {
      var date = new Date(2013, 3, 15);

      assert.equal(TimeFormat('%n %F %m %M', date), '3 April 04 Apr');
    });

    test('week', function() {
      var date = new Date(2013, 4, 15);

      assert.equal(TimeFormat('%w', date), '3');
    });

    test('year', function() {
      var date = new Date(2013, 4, 15);

      assert.equal(TimeFormat('%Y %y', date), '2013 13');
    });

    test('time', function() {
      var date = new Date(2013, 4, 15, 8, 30, 30, 50);

      assert.equal(TimeFormat('%g %G %h %h %i %s %u %a %A %e', date), '8 8 08 08 30 30 50 am AM ' + date.getTimezoneOffset());

      date = new Date(2013, 4, 15, 18, 5, 5, 5);

      assert.equal(TimeFormat('%g %G %h %H %i %s %u %a %A %e', date), '6 18 06 18 05 05 5 pm PM ' + date.getTimezoneOffset());
    });

    test('st,nd,rt,th', function() {
      var date = new Date(2013, 4, 1);

      assert.equal(TimeFormat('%S', date), 'st');

      date = new Date(2013, 4, 2);

      assert.equal(TimeFormat('%S', date), 'nd');

      date = new Date(2013, 4, 3);

      assert.equal(TimeFormat('%S', date), 'rd');

      date = new Date(2013, 4, 15);

      assert.equal(TimeFormat('%S', date), 'th');
    });

    test('midnight should be 12', function() {
      var date = new Date(2013, 4, 1);

      assert.equal(12, TimeFormat('%g', date));
    });
  });
});


test/dates.test.js
suite('dates', function() {

  suite('#getFirstDayOfWeek', function() {
    test('date', function() {
      var date = new Date(2013, 4, 15);

      var first = dateUtils.getFirstDayOfWeek(date);
      assert.equal(first.toDateString(), 'Sun May 12 2013');
    });
    test('date + offset', function() {
      var date = new Date(2013, 4, 15);

      var el = $('.weekly').weekly({
        currentDate: date,
        autoRender: false
      });

      var first = dateUtils.getFirstDayOfWeek(date, 1);
      assert.equal(first.toDateString(), 'Sun May 19 2013');
    });
  });

  suite('#getLastDayOfWeek', function() {
    test('date', function() {
      var date = new Date(2013, 4, 15);

      var first = dateUtils.getLastDayOfWeek(date);
      assert.equal(first.toDateString(), 'Sat May 18 2013');
    });
    test('date + offset', function() {
      var date = new Date(2013, 4, 15);

      var el = $('.weekly').weekly({
        currentDate: date,
        autoRender: false
      });

      var first = dateUtils.getLastDayOfWeek(date, 1);
      assert.equal(first.toDateString(), 'Sat May 25 2013');
    });
  });


  suite('#getWeekSpan', function() {

    test('same month', function() {
      var date = new Date(2013, 4, 15);

      var span = dateUtils.getWeekSpan(date);
      assert.equal(span, 'May 12 - 18');
    });

    test('different month', function() {
      var date = new Date(2013, 4, 29);

      var span = dateUtils.getWeekSpan(date);
      assert.equal(span, 'May 26 - Jun 01');
    });

    test('offset', function() {
      var date = new Date(2013, 4, 15);

      var span = dateUtils.getWeekSpan(date, 1);
      assert.equal(span, 'May 19 - 25');
    });

    test('day offset', function() {
      var date = new Date(2013, 4, 15);

      var span = dateUtils.getWeekSpan(date, 1, 1);
      assert.equal(span, 'May 20 - 26');
    });
  });

  suite('getDates', function() {

    test('dates should not have a time', function() {

      var date = new Date(2013, 4, 15, 5, 15);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].getHours(), 0);
      assert.equal(dates[0].getMinutes(), 0);
      assert.equal(dates[6].getHours(), 0);
      assert.equal(dates[6].getMinutes(), 0);

    });

    test('get dates in middle of month', function() {

      var date = new Date(2013, 4, 15);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun May 12 2013');
      assert.equal(dates[6].toDateString(), 'Sat May 18 2013');
    });

    test('get dates in beginning of month', function() {

      var date = new Date(2013, 4, 1);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun Apr 28 2013');
      assert.equal(dates[6].toDateString(), 'Sat May 04 2013');
    });


    test('get dates from a sunday', function() {

      var date = new Date(2013, 4, 5);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun May 05 2013');
      assert.equal(dates[6].toDateString(), 'Sat May 11 2013');
    });

    test('get dates from a sat', function() {

      var date = new Date(2013, 4, 11);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun May 05 2013');
      assert.equal(dates[6].toDateString(), 'Sat May 11 2013');
    });


    test('get dates end of month', function() {

      var date = new Date(2013, 4, 29);

      var dates = dateUtils.getDates(date, 0);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun May 26 2013');
      assert.equal(dates[6].toDateString(), 'Sat Jun 01 2013');
    });

    test('take into consideration weekOffset', function() {

      var date = new Date(2013, 4, 29);

      var dates = dateUtils.getDates(date, -1);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Sun May 19 2013');
      assert.equal(dates[6].toDateString(), 'Sat May 25 2013');

    });

    test('take into consideration dayOffset', function() {

      var date = new Date(2013, 4, 29);

      var dates = dateUtils.getDates(date, -1, 1);
      assert.equal(dates.length, 7);
      assert.equal(dates[0].toDateString(), 'Mon May 20 2013');
      assert.equal(dates[6].toDateString(), 'Sun May 26 2013');

    });
  });

  suite('getTimes', function() {

    test('get times', function() {
      var times = dateUtils.getTimes(8, 6);

      assert.equal(times.length, 10);
      assert.equal(times[0], '8:00 AM');
      assert.deepEqual(times, ['8:00 AM', '9:00 AM', '10:00 AM', '11:00 AM', '12:00 PM', '1:00 PM', '2:00 PM', '3:00 PM', '4:00 PM', '5:00 PM']);
    });

    test('get times overriding start and end', function() {
      var times = dateUtils.getTimes(6, 3);

      assert.deepEqual(times, ['6:00 AM', '7:00 AM', '8:00 AM', '9:00 AM', '10:00 AM', '11:00 AM', '12:00 PM', '1:00 PM', '2:00 PM']);

    });

    test('get times overriding start and end', function() {
      var times = dateUtils.getTimes(0, 3);

      assert.deepEqual(times, ['12:00 AM', '1:00 AM', '2:00 AM', '3:00 AM', '4:00 AM', '5:00 AM', '6:00 AM', '7:00 AM', '8:00 AM', '9:00 AM', '10:00 AM', '11:00 AM', '12:00 PM', '1:00 PM', '2:00 PM']);

    });

  });

  suite('realTimezoneOffset', function() {

    test('positive UTC offset', function() {
      var offset = +4;
      var real = offset - ((new Date()).getTimezoneOffset() / -60);

      assert.equal(real, dateUtils.realTimezoneOffset(offset));
    });

    test('negative UTC offset', function() {
      var offset = -4;
      var real = offset - ((new Date()).getTimezoneOffset() / -60);

      assert.equal(real, dateUtils.realTimezoneOffset(offset));
    });
  });

  suite('isPastDate', function(){
    test('past event should be true', function(){
      var pastDate = '2013-1-18';

      assert.equal(true, dateUtils.isPastDate(pastDate));
    });

    test('future event should be false', function(){
      var future = '2014-12-31';

      assert.equal(false, dateUtils.isPastDate(future));
    });
  });
});


lib/time.js
/**
 * Simple date and time formatter based on php's date() syntax.
 */

(function(w) {
  var oldRef = w.TimeFormat;

  var months = 'January|February|March|April|May|June|July|August|September|October|November|December'.split('|');
  var days = 'Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday'.split('|');

  var TimeFormat = function(format, time) {
    if(!time instanceof Date) return;

    // Implements PHP's date format syntax.
    return format.replace(/%d|%D|%j|%l|%S|%w|%F|%m|%M|%n|%Y|%y|%a|%A|%g|%G|%h|%H|%i|%s|%u|%e/g, function(match) {
      switch(match) {
        case '%d':
          return ("0" + time.getDate()).substr(-2,2);
        case '%D':
          return days[time.getDay()].substr(0,3);
        case '%j':
          return time.getDate();
        case '%l':
          return days[time.getDay()];
        case '%S':
          if(time.getDate() === 1) {
            return 'st';
          } else if(time.getDate() === 2) {
            return 'nd';
          } else if(time.getDate() === 3) {
            return 'rd';
          } else {
            return 'th';
          }
          break;
        case '%w':
          return time.getDay();
        case '%F':
          return months[time.getMonth()];
        case '%m':
          return ("0" + (time.getMonth() + 1)).substr(-2,2);
        case '%M':
          return months[time.getMonth()].substr(0,3);
        case '%n':
          return time.getMonth();
        case '%Y':
          return time.getFullYear();
        case '%y':
          return time.getFullYear().toString().substr(-2,2);
        case '%a':
          return time.getHours() > 11 ? 'pm' : 'am';
        case '%A':
          return time.getHours() > 11 ? 'PM' : 'AM';
        case '%g':
          return time.getHours() > 12 ? time.getHours() -12 : (time.getHours() ? time.getHours() : 12);
        case '%G':
        return time.getHours();
        case '%h':
          return ("0" + (time.getHours() > 12 ? time.getHours() -12 : time.getHours())).substr(-2,2);
        case '%H':
          return ("0" + time.getHours()).substr(-2,2);
        case '%i':
          return ("0" + time.getMinutes()).substr(-2,2);
        case '%s':
          return ("0" + time.getSeconds()).substr(-2,2);
        case '%u':
          return time.getMilliseconds();
        case '%e':
          return time.getTimezoneOffset();
      }
    });
  };

  TimeFormat.noConflict = function() {
    w.TimeFormat = oldRef;
    return TimeFormat;
  };

  w.TimeFormat = TimeFormat;

})(window);

lib/dates.js
/**
 * Date helpers
 */

(function(w){
  var oldRef = w.dateUtils;

  var dateUtils = {
    noConflict: function() {
      w.dateUtils = oldRef;
      return dateUtils;
    },
    getFirstDayOfWeek: function(date, offset) {
      offset = offset || 0;

      var first = date.getDate() - date.getDay();
      var newDate = new Date(date.getTime());
      newDate.setDate(first + (offset * 7));
      return newDate;
    },
    getLastDayOfWeek: function(date, weekOffset) {
      weekOffset = weekOffset || 0;

      var first = date.getDate() - date.getDay();
      var newDate = new Date(date.getTime());
      newDate.setDate(first + 6 + (weekOffset * 7));
      return newDate;
    },
    getDates: function(date, weekOffset, dayOffset) {
      date = new Date(date);
      date.setHours(0, 0, 0);
      dayOffset = dayOffset || 0;
      var daysInWeek = 7;

      var days = [];
      var sunday = this.getFirstDayOfWeek(date, weekOffset);

      for (var i = 0, c = daysInWeek; i < c; i++) {
        var d = new Date(sunday.getTime());
        d.setDate(d.getDate() - d.getDay() + dayOffset + i);
        days.push(d);
      }
      return days;
    },
    getTimes: function(startTime, endTime) {
      var end = endTime + 12;

      var times = [];
      for (var i = startTime; i < end; i++) {
        var hour = (i > 12 || i === 0) ? Math.abs(i - 12) : i;
        var timeString = hour+':00 ';

        timeString += (i > 11) ? 'PM' : 'AM';

        times.push(timeString);
      }

      return times;
    },
    getWeekSpan: function(date, offset, dayOffset) {
      dayOffset = dayOffset || 0;
      var first = this.getFirstDayOfWeek(date, offset);
      var last = this.getLastDayOfWeek(date, offset);

      first.setDate(first.getDate() + dayOffset);
      last.setDate(last.getDate() + dayOffset);

      var span = TimeFormat('%M %d', first) + ' - ';
      if (first.getMonth() == last.getMonth()) {
        span += TimeFormat('%d', last);
      } else {
        span += TimeFormat('%M %d', last);
      }
      return span;
    },
    realTimezoneOffset: function(offset) {
      var local = (new Date()).getTimezoneOffset() / -60;
      var real = offset - local;
      return real;
    },
    isPastDate: function(past) {
      var pastParts = past.split('-');
      return (TimeFormat('%Y%m%d', new Date(pastParts[0], pastParts[1], pastParts[2])) < TimeFormat('%Y%m%d', new Date()));
    }
  };

  w.dateUtils = dateUtils;
})(window);


lib/weekly.js
(function($) {

  $.declare('weekly', {
    defaults: {
      startTime: 0,
      endTime: 12,
      startTimeScrollOffset: '8:00 AM',
      weekOffset: 0,
      currentDate: new Date(),
      autoRender: true,
      fitText: true,
      fitTextMin: 12,
      fitTextMax: 15,
      template: '##template##',
      readOnly: false,
      enableResize: true,
      enableDelete: true,
      autoSplit: false,
      autoSplitInterval: 30,
      showToday: true,
      allowPreviousWeeks: true,
      allowPastEventCreation: false,
      timezoneOffset: 0,
      utcOffset: ((new Date()).getTimezoneOffset() / -60),
      todayFirst: false,
      dayOffset: 0,

      // How many minutes to draw a divider line
      interval: 30
    },

    events: {
      'click .weekly-event': 'eventClicked'
    },

    init: function() {
      this.events = [];

      this.oldDate = this.currentDate;

      if (this.readOnly) {
        this.enableResize = false;
        this.enableDelete = false;
      }

      if (this.todayFirst) {
        this.dayOffset = this.currentDate.getDay();
      }

      if(this.interval < 1 || this.interval > 60) {
        this.interval = 60;
      }

      if (this.autoRender) {
        var data = this.update();
        this.emit('weekChange', data);
      }
    },

    get: function(property) {
      return this[property];
    },

    update: function() {

      var data = {
        timef: TimeFormat,
        getWeekSpan: dateUtils.getWeekSpan,
        currentDate: this.currentDate,
        dates: dateUtils.getDates(this.currentDate, this.weekOffset, this.dayOffset),
        times: dateUtils.getTimes(this.startTime, this.endTime),
        showPreviousWeekButton: (this.allowPreviousWeeks || (this.weekOffset !== 0))
      };
      this.render(data);

      for(var i = 0, c = this.events.length; i < c; i++) {
        this.renderEvent(this.events[i]);
      }

      this.timeDifference = (this.endTime + 12) - this.startTime;

      this.registerClickToCreate();
      this.registerModifyEvent();

      this.highlightToday();

      if(!this.showToday || !this.weekOffset) {
        this.el.find(".weekly-change-today-button").css('display', 'none');
      } else {
        this.el.find(".weekly-change-today-button").css('display', 'block');
      }

      this.el.find('.weekly-time-navigation .weekly-previous-week .week').html(dateUtils.getWeekSpan(this.currentDate, this.weekOffset - 1, this.dayOffset));
      this.el.find('.weekly-time-navigation .weekly-next-week .week').html(dateUtils.getWeekSpan(this.currentDate, this.weekOffset + 1, this.dayOffset));

      this.el.find('.weekly-time-navigation .weekly-header').html(dateUtils.getWeekSpan(this.currentDate, this.weekOffset, this.dayOffset));

      if (this.fitText) {
        this.el.find(".weekly-days .weekly-day, .weekly-times .weekly-time").fitText(1, {
          'minFontSize': this.fitTextMin,
          'maxFontSize': this.fitTextMax
        });
      }

      if (this.startTimeScrollOffset) {
        var top = $(window).scrollTop();
        var el = this.el.find('[data-time="'+this.startTimeScrollOffset+'"]');
        el[0].scrollIntoView();
        $(window).scrollTop(top);
      }

      return { dates: data.dates, times: data.times };
    },

    highlightToday: function() {
      var today = this.currentDate;

      this.el.find('.weekly-grid [data-date="' + TimeFormat('%Y-%n-%j', today) + '"]').addClass('weekly-today');
    },

    registerClickToCreate: function() {
      this.mouseDown = false;
      this.pendingEvent = null;
      this.pendingEventStart = null;

      var gridDays = this.el.find('.weekly-grid .weekly-day');

      // Make sure anything previously bound is bound no more.
      gridDays.unbind('mousedown mousemove mouseup mouseout click');

      if (this.readOnly) {
        return;
      }

      gridDays.on('mousedown', this.proxy(function(event){
        var target = $(event.target);

        if(event.which !== 1 || target.is('.weekly-dragger') || target.is('.weekly-delete')) return;

        var currentTarget = $(event.currentTarget);

        if(!this.allowPastEventCreation && dateUtils.isPastDate(currentTarget.data('date'))) {
          return;
        }

        this.mouseDown = true;

        if(this.pendingEvent) {
          this.pendingEvent.remove();
          this.pendingEvent = null;
        }
      }));

      gridDays.on('mouseup', this.proxy(function(){
        if(!this.mouseDown) return;

        this.mouseDown = false;

        if(this.pendingEvent) {
          var eventData = this.pendingEvent.data();

          var parsedDate = eventData.date.split('-');

          this.addEvent({
            start: new Date(parsedDate[0], parsedDate[1], parsedDate[2], ~~(eventData.starttime) - this.timezoneOffset, this.fromDecimal(eventData.starttime)),
            end: new Date(parsedDate[0], parsedDate[1], parsedDate[2], ~~(eventData.endtime) - this.timezoneOffset, this.fromDecimal(eventData.endtime))
          });

          this.pendingEvent.remove();
          this.pendingEvent = null;
          this.pendingEventStart = null;
        }
      }));

      gridDays.on('mousemove', this.proxy(function(event){
        if(this.mouseDown) {
          this.createEvent(event);
        }
      }));

      gridDays.on('click', this.proxy(function(event){
        var target = $(event.currentTarget);

        if(!this.allowPastEventCreation && dateUtils.isPastDate(target.data('date'))) {
          return;
        }

        if($(event.target).is('.weekly-time,.weekly-day')) {
          this.mouseDown = true;
          this.createEvent(event);
          gridDays.trigger('mouseup');
        }
      }));

      gridDays.on('mouseleave', this.proxy(function(event){
        if(this.mouseDown) {
          gridDays.trigger('mouseup');
        }
      }));
    },

    registerModifyEvent: function() {
      this.mouseModifyDown = false;

      var eventDraggers = this.el.find('.weekly-grid');

      this.currentDragger = null;

      // Make sure anything previously bound is bound no more.
      eventDraggers.find('.weekly-dragger').unbind('mousedown mousemove mouseup mouseout click');

      if (this.readOnly) {
        return;
      }

      eventDraggers.on('mousedown', '.weekly-dragger', this.proxy(function(event){
        if(event.which !== 1) return;
        this.mouseModifyDown = true;

        this.currentDragger = $(event.target);

        this.eventOffset = 0;
      }));

      eventDraggers.on('mouseup', '.weekly-day', this.proxy(function(){
        this.mouseModifyDown = false;

        this.currentDragger = null;
      }));

      eventDraggers.on('mousemove', '.weekly-day', this.proxy(function(event){
        if(this.mouseModifyDown) {
          this.modifyEvent(event);
        }
      }));

      eventDraggers.on('mouseleave', this.proxy(function(event){
        if(this.mouseModifyDown) {
          this.currentDragger.trigger('mouseup');
        }
      }));
    },

    createEvent: function(event) {
      var target = $(event.currentTarget);
      var targetOffset = target.parent().offset();
      var mouseOffsetTop = event.pageY - targetOffset.top;
      var dayHeight = $(event.currentTarget).height();
      var hourHeight = Math.round(dayHeight / this.timeDifference);
      var intervalHeight = hourHeight / (60 / this.interval);

      var tempStart = Math.floor(mouseOffsetTop / intervalHeight) * intervalHeight;
      var tempEnd = Math.ceil(mouseOffsetTop / intervalHeight) * intervalHeight;

      if(this.pendingEventStart === null) {
        this.pendingEventStart = tempStart;
      }

      if(this.pendingEventEnd === null || tempEnd > this.pendingEventStart) {
        this.pendingEventEnd = tempEnd;
      }

      if(!this.pendingEvent) {
        target.append('<div class="weekly-event-pending"></div>');
        this.pendingEvent = target.find('.weekly-event-pending');
        this.pendingEvent.data('date', target.data('date'));
      }

      this.pendingEvent.css({
        top: this.pendingEventStart,
        bottom: dayHeight - this.pendingEventEnd
      });

      this.pendingEvent.data('starttime', ((this.pendingEventStart / hourHeight) || 0) + this.startTime);
      this.pendingEvent.data('endtime', ((this.pendingEventEnd / hourHeight) || 1) + this.startTime);
    },

    modifyEvent: function(event) {
      var target = this.currentDragger.parents('.weekly-event');
      var targetOffset = target.parent().offset();
      var mouseOffsetTop = event.pageY - targetOffset.top;
      var dayHeight = $(event.currentTarget).height();
      var hourHeight = Math.round(dayHeight / this.timeDifference);
      var intervalHeight = hourHeight / (60 / this.interval);

      var tempEnd = Math.ceil(mouseOffsetTop / intervalHeight) * intervalHeight;

      if(tempEnd < (targetOffset.top + dayHeight)) {
        target.css({
          bottom: dayHeight - tempEnd
        });
      }

      var duration = target.outerHeight() / hourHeight;
      var end = new Date(target.data('start'));
      end.setHours(end.getHours() + ~~(duration));
      end.setMinutes(this.fromDecimal(duration));
      target.data('end', end);

      this.events[target.data('_index')].end = end;

      this.el.trigger('modifyEvent', this.events[target.data('_index')]);
    },

    nextWeek: function() {
      this.changeDate(1);
    },

    prevWeek: function() {
      this.changeDate(-1);
    },

    jumpToday: function() {
      this.changeDate(0);
    },

    changeDate: function(offsetWeek) {
      if(!offsetWeek) {
        this.weekOffset = 0;
      } else {
        this.weekOffset += offsetWeek;
      }

      var data = this.update();
      this.emit('weekChange', data);
    },

    renderEvent: function(event) {
      var start = new Date(event.start.getTime());
      var end = new Date(event.end.getTime());

      start.setHours(start.getHours() + this.timezoneOffset);
      end.setHours(end.getHours() + this.timezoneOffset);

      var startDate = start.getFullYear() + "-" + start.getMonth() + "-" + start.getDate();
      var startTime = start.toTimeString().slice(0,5);
      var endDate = end.getFullYear() + "-" + event.end.getMonth() + "-" + end.getDate();
      var endTime = end.toTimeString().slice(0,5);

      if(endTime === "00:00") {
        endTime = "24:00";
      }

      var topOffset = 100 - this.getTimeOffsetPercent(startTime);
      var bottomOffset = this.getTimeOffsetPercent(endTime);

      var eventTemplate = $('<div class="weekly-event"></div>');

      eventTemplate.data(event);

      eventTemplate.data('offset-start', start);
      eventTemplate.data('offset-end', end);

      eventTemplate.css({
        top: topOffset + '%',
        bottom: bottomOffset + '%'
      }).append([
        '<button data-action="removeEvent" class="weekly-delete">&times;</button>',
        '<div class="weekly-event-time">' + TimeFormat('%g:%i', start) + ' - ' + TimeFormat('%g:%i%a', end) + '</div>',
        '<div class="weekly-event-title">' + event.title + '</div>',
        '<div class="weekly-event-desc">' + event.description + '</div>',
        '<div class="weekly-dragger"></div>'
      ].join(''));

      if (!this.enableResize) {
        eventTemplate.find('.weekly-dragger').remove();
      }

      if (!this.enableDelete) {
        eventTemplate.find('.weekly-delete').remove();
      }

      if (event.type) {
        eventTemplate.addClass('weekly-event-'+event.type.replace(/ /g, '-'));
      }

      var selectedDay = this.el.find('.weekly-grid .weekly-day[data-date="' + startDate + '"]');

      selectedDay.append(eventTemplate);

      if (this.fitText) {
        selectedDay.find(".weekly-event-title").fitText(1, {
          'minFontSize': this.fitTextMin,
          'maxFontSize': this.fitTextMax
        });
      }
    },

    toFraction: function(time) {
      if(time.toString().indexOf(':') === -1) {
        return time;
      }

      var parts = time.toString().split(':');

      return parseFloat(parts[0] + '.' + 100/(60/parts[1]));
    },

    // This just gets the minutes in the decimal
    fromDecimal: function(time) {
      return Math.round(60 * (time % 1));
    },

    getTimeOffsetPercent: function(time) {
      time = this.toFraction(time) - this.startTime;

      var diff = this.timeDifference - time;

      if(diff < 0) {
        diff = time - this.timeDifference;
      }

      var percent = (diff / this.timeDifference) * 100;

      return percent;
    },

    setAutoSplit: function(val) {
      this.autoSplit = val;
    },

    setSplitInterval: function(val) {
      this.autoSplitInterval = val;
    },

    splitEvent: function(event) {
      var diff = event.end.getTime() - event.start.getTime();
      var interval = this.autoSplitInterval * 60 * 1000;
      var count = Math.ceil(diff / interval); //divide by 1 hour
      var startTime = event.start.getTime();
      var events = [];
      for (var i = 0; i < count; i++) {
        var newEvent = $.extend({}, event); //clone event
        newEvent.start = new Date(startTime + (interval * i));
        newEvent.end = new Date(startTime + (interval * (i+1)));
        events.push(newEvent);
      }
      return events;
    },

    addEvent: function(event) {
      if(event instanceof Array) {
        for(var i = 0, c = event.length; i < c; i++) {
          this.addEvent(event[i]);
        }
        return;
      }

      if (this.autoSplit) {
        event = this.splitEvent(event);
      } else {
        event = [event];
      }

      for (var x = 0, y = event.length; x < y; x++) {
        var e = event[x];

        e = $.extend({
          title: '',
          description: '',
          start: null,
          end: null
        }, e);

        e._index = this.events.length;

        if (e.start.getHours() >= this.startTime && e.end.getHours() <= (this.endTime + 12)) {
          this.renderEvent(e);
        }
        this.events.push(e);
      }

      var eventData = (this.autoSplit) ? event : event[0];
      this.emit('addEvent', [eventData]);
    },

    removeEvent: function(e) {
      var el = $(e.target).parents('.weekly-event');
      var event = el.data();

      this.events.splice(event._index, 1);
      el.remove();

      this.el.trigger('removeEvent', event);

      return false;
    },

    clearEvents: function() {
      this.el.find('.weekly-event').remove();

      this.events = [];

      this.el.trigger('clearEvents');
    },

    eventClicked: function(e) {
      var el = $(e.currentTarget);
      var event = el.data();
      this.emit('eventClick', [event, el]);
    },

    setTimezoneOffset: function(offset) {
      this.timezoneOffset = dateUtils.realTimezoneOffset(offset);

      this.utcOffset = offset;

      this.update();

      return this;
    },

    setReadOnly: function(val) {
      this.readOnly = val;
      this.enableResize = !val;
      this.enableDelete = !val;
      this.update();
      return this;

    }

  });

})(jQuery);
